akka-projection-kafka/src/main/scala/akka/projection/kafka/KafkaSourceProvider.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.kafka
+
+import scala.concurrent.Future
+
+import akka.annotation.InternalApi
+import akka.kafka.AutoSubscription
+import akka.kafka.ConsumerSettings
+import akka.kafka.scaladsl.Consumer
+import akka.projection.internal.MergeableOffsets
+import akka.projection.internal.MergeableOffsets.OffsetRow
+import akka.projection.scaladsl.SourceProvider
+import akka.stream.scaladsl.Source
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.TopicPartition
+
+object KafkaSourceProvider {
+  private val regexTp = """(.+)-(\d+)""".r
+
+  /**
+   * Create a [[SourceProvider]] that resumes from externally managed offsets
+   */
+  def apply[Key, ValueEnvelope](
+      settings: ConsumerSettings[Key, ValueEnvelope],
+      subscription: AutoSubscription): SourceProvider[MergeableOffsets.Offset, ConsumerRecord[Key, ValueEnvelope]] =
+    new KafkaOffsetSourceProvider[Key, ValueEnvelope](settings, subscription)
+
+  @InternalApi
+  class KafkaOffsetSourceProvider[Key, ValueEnvelope] private[kafka] (
+      settings: ConsumerSettings[Key, ValueEnvelope],
+      // all Alpakka Source's need a subscription, but we could generate a topic subscription automatically based on the provided offset.
+      // if the offset is None though we're dead in the water
+      subscription: AutoSubscription)
+      extends SourceProvider[MergeableOffsets.Offset, ConsumerRecord[Key, ValueEnvelope]] {
+
+    override def source(offsets: Option[MergeableOffsets.Offset]): Source[ConsumerRecord[Key, ValueEnvelope], _] = {
+      offsets match {
+        case Some(mergeableOffsets) =>
+          val getOffsetsOnAssign = (tps: Set[TopicPartition]) =>
```

Comment:
The `getOffsetsOnAssign` is called "When a topic-partition is assigned to a consumer". That can occur later after startup and at that point the `mergeableOffsets` from the `offsets` parameters are stale. They could have been updated by other projection instance after we read it at startup. Meaning that this `getOffsetsOnAssign` must read from the offset store when it's called.

One way of doing that and still keep the `offsets` parameter is that the `MergeableOffsets.Offset` is something that can be queried for the full Map, called from here. Since it's created by the OffsetStore it can have a reference to the OffsetStore, which performs the actual query.

Reply:
Good point.  Maybe we could pass in a thunk to the `SourceProvider` to delegate calling `readOffset`. In both implementations we don't need access to the offset except to pass to the `SourceProvider` itself.

Reply:
I think the OffsetStore abstraction adds too much complexity, since it has to be different in types for save offset for different impl (future vs dbio).

Reply:
Ok, I've used a thunk instead.